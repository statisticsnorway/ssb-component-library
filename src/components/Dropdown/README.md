Dropdown
========

> A component that allows a user to select a value from a series of options.

### Usage

#### HTML
This component needs some javascript functionality to work.
```html
<div class="ssb-dropdown">
    <label for="input-dropdown">Menu header</label>
    <div class="dropdown-interactive-area">
        <button onclick="toggle the open class">
            <input class="focused" id="input-dropdown" disabled="" placeholder="Select item" value="">
            <!-- If closed -->
            <i class="dd-icon">{feather.chevronDown 18px}</i>
            <!-- If open -->
            <i class="dd-icon">{feather.chevronUp 18px}</i>
        </button>
        <div class="list-of-options">
            <button class="option-list-element" value="item1" id="item1" onclick="select item and close">Apples</button>
            <button class="option-list-element" value="item2" id="item2" onclick="select item and close">Rainbows</button>
            <button class="option-list-element" value="item3" id="item3" onclick="select item and close">Ocean</button>
        </div>
    </div>
</div>
```

#### React

```jsx harmony
const items = [
	{
		title: 'Apples',
		id: 'item1',
	}, {
		title: 'Rainbows',
		id: 'item2',
	}, {
		title: 'Ocean',
		id: 'item3',
	}, {
		title: 'Automobiles and other fun things to talk about',
		id: 'item4',
	}, 
];

<Dropdown header="Menu header" items={items} />

<Dropdown header="Menu header" selectedItem={{ title: 'Ocean', id: 'item3' }} items={items} />

<Dropdown header="Menu header" searchable items={items} />
```

Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| header | string | Displays header |
| items | arrayOf(title, id) |Required items for rendering dropdown |
| onSelect | func | Callback function when a title is selected |
| open | bool | Selected state of dropdown |
| placeholder | string | Placeholder for dropdown |
| searchable | bool | If true, dropdown is searchable |
| selectedItem | object | Selected element in dropdown |
