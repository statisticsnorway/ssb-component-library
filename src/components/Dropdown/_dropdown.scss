.ssb-dropdown {
  @include roboto;
  display: flex;
  flex-direction: column;
  transition: all .25s ease-in-out;
  max-width: 350px;
  min-width: 150px;

  label {
    font-size: 14px;
    line-height: normal;
    margin-bottom: 5px;
  }

  .dropdown-interactive-area {
    @include focus-marker;
    @include reset-button;
    cursor: pointer;
    position: relative;

    .dd-icon {
      color: $ssb-green-4;
      font-size: 24px;
      position: absolute;
      right: 12px;
      top: 10px;
    }

    input {
      background-color: $ssb-white;
      border: 1px solid $ssb-dark-6;
      box-sizing: border-box;
      cursor: pointer;
      font-size: 16px;
      height: $input-field-height;
      padding: 4px $input-field-height 4px 12px;
      text-overflow: ellipsis;
      width: 100%;

      &::placeholder {
        color: $ssb-dark-6;
      }

      &:hover {
        border: 1px solid $ssb-green-4;
      }

      &:focus, &.focused {
        border: 1px solid $ssb-green-4;
        outline: $ssb-green-4 auto 1px;
      }
    }

    .list-of-options {
      @include roboto;
      background: $ssb-white;
      border: 1px solid $ssb-dark-6;
      left: 0;
      list-style: none;
      margin: 0;
      max-height: 235px;
      min-width: 150px;
      overflow-y: auto;
      padding-left: 0;
      position: absolute;
      scrollbar-color: $scrollbar-color $scrollbar-color; // Firefox experimental property
      scrollbar-width: thin;
      top: $input-field-height;
      z-index: 9999;
      width: 100%;

      .option-list-element {
        @include focus-marker;
        @include reset-button;
        @include roboto;
        background: $ssb-white;
        cursor: pointer;
        display: block;
        font-size: 16px;
        font-weight: normal;
        line-height: 1.25;
        overflow: hidden;
        padding: 12px;
        text-overflow: ellipsis;
        transition: all .18s;
        width: 100%;

        &:hover, &:focus {
          background: $ssb-dark-5;
          color: $ssb-white;
          z-index: 2;
        }

        &.selected {
          background: $ssb-green-4;
          color: $ssb-white;
          font-weight: bold;
        }
      }

    }

    ::-webkit-scrollbar {
      width: 6px;
    }

    ::-webkit-scrollbar-track {
      background: $ssb-white;
      border-bottom: 1px solid $ssb-dark-6;
    }

    ::-webkit-scrollbar-thumb {
      background: $scrollbar-color;
      border-radius: 3px;
    }
  }
}
