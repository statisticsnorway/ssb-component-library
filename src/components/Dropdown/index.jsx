import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import uuid from 'uuid/v4';

const Dropdown = ({
	header, icon, items, onSelect, open, placeholder, searchable, selectedItem, tabIndex,
}) => {
	const id = uuid();
	const node = useRef();
	const [isOpen, setOpen] = useState(open);
	const [availableOptions, filterOptions] = useState(items);
	const [selectedOption, selectItem] = useState(selectedItem || { title: '', id: '' });
	const [inputFieldValue, updateInputValue] = useState('');

	const filterItems = event => {
		updateInputValue(event.target.value);
		filterOptions(items.filter(it => it.title.toLowerCase().includes(event.target.value.toLowerCase())));
	};

	const handleSelection = item => {
		selectItem({ title: item.title, id: item.id });
		onSelect(item);
		setOpen(false);
		filterOptions(items);
		updateInputValue('');
	};

	const handleClickOutside = e => {
		if (!node.current.contains(e.target)) {
			setOpen(false);
		}
	};

	useEffect(() => {
		if (isOpen) {
			document.addEventListener('mousedown', handleClickOutside);
		} else {
			document.removeEventListener('mousedown', handleClickOutside);
		}
		return () => {
			document.removeEventListener('mousedown', handleClickOutside);
		};
	}, [isOpen]);

	return (
		<div className="ssb-dropdown">
			{header && <label htmlFor={id}>{header}</label>}
			<button className="dropdown-interactive-area" tabIndex={tabIndex} ref={node} onClick={() => setOpen(!isOpen)}>
				<input
					className={isOpen ? 'focused' : ''}
					id={id}
					onChange={e => filterItems(e)}
					disabled={!searchable}
					placeholder={selectedOption.title ? selectedOption.title : placeholder}
					value={inputFieldValue}
				/>
				<div className="dd-icon">{icon}</div>
				{isOpen && (
					<div className="list-of-options">
						{availableOptions.map(it => (
							<button
								className={`option-list-element${selectedOption.id === it.id ? ' selected' : ''}`}
								key={it.id}
								onClick={() => handleSelection(it)}
								value={it.id}
								id={it.id}
							>{it.title}
							</button>
						))}
					</div>
				)}
			</button>
		</div>
	);
};

Dropdown.defaultProps = {
	header: '',
	items: [{ id: '', title: '' }],
	onSelect: () => {},
	open: false,
	searchable: false,
	placeholder: '-- Select --',
};

Dropdown.propTypes = {
	header: PropTypes.string,
	icon: PropTypes.node,
	items: PropTypes.arrayOf(PropTypes.shape({
		title: PropTypes.string,
		id: PropTypes.string,
	})),
	onSelect: PropTypes.func,
	open: PropTypes.bool,
	placeholder: PropTypes.string,
	searchable: PropTypes.bool,
	selectedItem: PropTypes.object,
	tabIndex: PropTypes.number,
};

export default Dropdown;
