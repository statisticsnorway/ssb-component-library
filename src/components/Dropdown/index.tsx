import React, { useEffect, useRef, useState, type ReactNode } from 'react'
import { ChevronDown, ChevronUp } from 'react-feather'
import { useId } from '../../utils/useId'
import InputError from '../InputError'
import classNames from '../../utils/utils'

interface Item {
  id: string
  title: string
  disabled?: boolean
}

interface DropdownProps {
  className?: string
  header?: string
  icon?: ReactNode
  items?: Item[]
  onSelect?: (item: Item) => void
  open?: boolean
  placeholder?: string
  searchable?: boolean
  selectedItem?: Item
  tabIndex?: number
  error?: boolean
  errorMessage?: string
  ariaLabel?: string
  id?: string
  largeSize?: boolean
}

const Dropdown: React.FC<DropdownProps> = ({
  className,
  header = '',
  icon,
  items = [{ id: '', title: '' }],
  onSelect = () => {},
  open = false,
  placeholder = '-- Select --',
  searchable = false,
  selectedItem,
  tabIndex,
  error,
  errorMessage,
  ariaLabel,
  id,
  largeSize,
}) => {
  const wrapper = useRef<HTMLDivElement>(null)
  const node = useRef<HTMLButtonElement>(null)
  const itemRefs = useRef<{ [key: number]: HTMLLIElement | null }>({})
  const optionList = useRef<HTMLUListElement>(null)

  const [isOpen, setIsOpen] = useState(open)
  const [availableOptions, setAvailableOptions] = useState(items || [])
  const DEFAULT_ITEM = { title: '', id: '' }
  const [selectedOption, setSelectedOption] = useState(selectedItem || DEFAULT_ITEM)
  const [activeOption, setActiveOption] = useState(selectedItem || DEFAULT_ITEM)
  const [inputFieldValue, setInputFieldValue] = useState('')
  const [keyNavPosition, setKeyNavPosition] = useState(0)

  const dropdownId = id ?? useId()

  useEffect(() => {
    setSelectedOption(selectedItem || DEFAULT_ITEM)
    setActiveOption(selectedItem || DEFAULT_ITEM)
  }, [selectedItem])

  const filterOptions = (value: string) => {
    setInputFieldValue(value)
    setAvailableOptions(items.filter((it) => it.title.toLowerCase().includes(value.toLowerCase())))
  }

  useEffect(() => {
    setAvailableOptions(items)
  }, [items])

  const filterItems = (event: React.ChangeEvent<HTMLInputElement>) => {
    filterOptions(event.target.value)
  }

  const handleSelection = (item: Item) => {
    if (!item.disabled) {
      setSelectedOption({ title: item.title, id: item.id })
      onSelect(item)
      setIsOpen(false)
    }

    if (items[keyNavPosition].id !== item.id) {
      const idx = items.findIndex((it) => it.id === item.id)
      setKeyNavPosition(idx)
    }

    if (searchable) {
      setAvailableOptions(items)
      setInputFieldValue('')
    }

    if (node.current) {
      node.current.focus()
    }
  }

  const handleClickOutside = (e: MouseEvent) => {
    if (wrapper.current && !wrapper.current.contains(e.target as Node)) {
      setTimeout(() => {
        setIsOpen(false)
        if (searchable) {
          setInputFieldValue('')
        }
      }, 100)
    }
  }

  const handleKeyboardNav = (e: React.KeyboardEvent) => {
    if (e.key === 'ArrowUp' && keyNavPosition > 0) {
      e.preventDefault()
      setKeyNavPosition(keyNavPosition - 1)
    } else if (e.key === 'ArrowDown' && keyNavPosition < items.length - 1) {
      e.preventDefault()
      setKeyNavPosition(keyNavPosition + 1)
    } else if (e.key === 'Enter') {
      e.preventDefault()
      if (isOpen) {
        handleSelection(items[keyNavPosition])
      } else {
        setIsOpen(true)
      }
    } else if (e.key === 'Escape') {
      if (!searchable) {
        e.preventDefault()
        setIsOpen(false)
      }
    }
  }

  const handleSearchSpecialKeys = (e: React.KeyboardEvent) => {
    if (e.key === 'Escape') {
      setIsOpen(false)
    } else if (e.key === 'Tab') {
      setIsOpen(false)
    } else if (e.key === 'ArrowDown' && keyNavPosition < availableOptions.length - 1) {
      setKeyNavPosition(keyNavPosition + 1)
    } else if (e.key === 'ArrowUp' && keyNavPosition > 0) {
      setKeyNavPosition(keyNavPosition - 1)
    } else if (e.key === 'Enter' && isOpen) {
      e.preventDefault()
      handleSelection(availableOptions[keyNavPosition])
    } else {
      setIsOpen(true)
      if (e.key === ' ') {
        if (searchable) {
          e.preventDefault()
          filterOptions(`${(e.currentTarget as HTMLInputElement).value} `)
        }
      } else {
        filterOptions((e.currentTarget as HTMLInputElement).value)
      }
    }
  }

  useEffect(() => {
    if (isOpen && itemRefs.current[keyNavPosition]) {
      itemRefs.current[keyNavPosition]?.scrollIntoView({
        behavior: 'smooth',
        block: 'nearest',
        inline: 'start',
      })
      setActiveOption(items[keyNavPosition])
    }
  }, [keyNavPosition])

  useEffect(() => {
    if (isOpen) {
      if (!searchable) {
        optionList.current?.focus()
      }
      document.addEventListener('mousedown', handleClickOutside)
    }

    if (isOpen && keyNavPosition === 0) {
      setActiveOption(items[0])
    }

    return () => {
      document.removeEventListener('mousedown', handleClickOutside)
    }
  }, [isOpen])

  useEffect(() => {
    if (searchable && !inputFieldValue) {
      setKeyNavPosition(0)
    }
  }, [inputFieldValue])

  const renderIcon = () => {
    if (icon) {
      return <div className='dd-icon'>{icon}</div>
    }
    return isOpen ? (
      <ChevronUp className='dd-icon' size={largeSize ? 50 : 24} />
    ) : (
      <ChevronDown className='dd-icon' size={largeSize ? 50 : 24} />
    )
  }

  const renderHeader = () => {
    if (!header && ariaLabel) {
      return (
        <span className='dropdown-label sr-only' id={`label_${dropdownId}`}>
          {ariaLabel}
        </span>
      )
    }
    return (
      <span id={`label_${dropdownId}`} className='dropdown-label'>
        {header}
      </span>
    )
  }
  const renderHeaderSearchable = () => header && <label htmlFor={`input_${dropdownId}`}>{header}</label>

  return (
    <div
      id={`dropdown_${dropdownId}`}
      className={classNames('ssb-dropdown', className, error && ' error', largeSize && ' large')}
    >
      {searchable ? renderHeaderSearchable() : renderHeader()}
      <div className='dropdown-interactive-area' ref={wrapper} tabIndex={tabIndex}>
        {!searchable && (
          <button
            className={isOpen ? 'focused opener' : 'opener'}
            id={`button_${dropdownId}`}
            ref={node}
            tabIndex={0}
            onClick={() => {
              setIsOpen(!isOpen)
            }}
            onKeyDown={(e) => {
              handleKeyboardNav(e)
            }}
            type='button'
            aria-expanded={isOpen ? 'true' : 'false'}
            aria-describedby={error && errorMessage ? `error_${dropdownId}` : undefined}
            aria-haspopup='listbox'
            aria-labelledby={
              !header && !ariaLabel ? `button_${dropdownId}` : `label_${dropdownId} button_${dropdownId}`
            }
          >
            {selectedOption?.title ? selectedOption.title : placeholder}
          </button>
        )}
        {searchable && (
          <input
            className={isOpen ? 'focused' : ''}
            id={`input_${dropdownId}`}
            onKeyDown={handleSearchSpecialKeys}
            onChange={filterItems}
            onFocus={() => setIsOpen(!isOpen)}
            disabled={!searchable}
            placeholder={selectedOption.title ? selectedOption.title : placeholder}
            value={inputFieldValue}
            aria-describedby={error && errorMessage ? `error_${dropdownId}` : undefined}
            role='combobox'
            aria-autocomplete='list'
            aria-expanded={isOpen ? 'true' : 'false'}
            aria-controls={`list_of_options_${dropdownId}`}
            aria-label={ariaLabel}
            type='text'
            aria-activedescendant={activeOption ? activeOption.id : undefined}
          />
        )}
        {renderIcon()}
        {isOpen && (
          <ul
            id={`list_of_options_${dropdownId}`}
            className={`list-of-options${!isOpen ? ' hidden' : ''}`}
            role='listbox'
            aria-labelledby={!searchable && (header || ariaLabel) ? 'dropdown-label' : undefined}
            aria-activedescendant={!searchable && activeOption.id !== '' ? activeOption.id : undefined}
            tabIndex={-1}
            ref={optionList}
            onKeyDown={(e) => {
              handleKeyboardNav(e)
            }}
          >
            {availableOptions.map((it, idx) => (
              <li
                key={it.id}
                className={`option-list-element${selectedOption?.id === it.id ? ' selected' : ''}${isOpen && keyNavPosition === idx ? ' active' : ''}${it.disabled ? ' disabled' : ''}`}
                onClick={() => {
                  handleSelection(it)
                }}
                id={it.id}
                ref={(el) => {
                  itemRefs.current[idx] = el
                }}
                role='option'
                aria-selected={isOpen && keyNavPosition === idx ? 'true' : undefined}
                aria-disabled={it.disabled}
              >
                {it.title}
              </li>
            ))}
          </ul>
        )}
        {error && errorMessage && <InputError errorMessage={errorMessage} id={`error_${dropdownId}`} />}
      </div>
    </div>
  )
}

export default Dropdown
