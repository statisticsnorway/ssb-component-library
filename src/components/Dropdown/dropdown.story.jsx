import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import { ChevronDown } from 'react-feather';
import Dropdown from './index';

const items = [
	{
		title: 'Apples',
		id: 'item1',
	}, {
		title: 'Rainbows',
		id: 'item2',
	}, {
		title: 'Ocean',
		id: 'item3',
	}, {
		title: 'Automobiles and other fun things to talk about',
		id: 'item4',
	}, {
		title: 'Banan',
		id: 'item5',
	}, {
		title: 'Appelsin',
		id: 'item6',
	}, {
		title: 'Fersken',
		id: 'item7',
	}, {
		title: 'Mandarin',
		id: 'item8',
	},
];

storiesOf('Dropdown', module).addDecorator(centered)
	.add('Standard', () => (
		<Dropdown header="Menu header" icon={<ChevronDown />} items={items} />
	))
	.add('Pre-selected value', () => (
		<Dropdown header="Menu header" icon={<ChevronDown />} selectedItem={{ title: 'Ocean', id: 'item3' }} items={items} />
	))
	.add('Searchable', () => (
		<Dropdown header="Menu header" icon={<ChevronDown />} searchable items={items} placeholder="Search and select" />
	));
