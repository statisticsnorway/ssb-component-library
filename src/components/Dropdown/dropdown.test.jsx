import React from 'react';
import {mount, shallow} from 'enzyme';
import Dropdown from './index';

const items = [
	{
		title: 'Apples',
		id: 'item1',
	}, {
		title: 'Rainbows',
		id: 'item2',
	}, {
		title: 'Ocean',
		id: 'item3',
	},
];

jest.mock('uuid/v4', () => jest.fn(() => 1));

describe('Dropdown component', () => {
	test('Matches the snapshot', () => {
		const wrapper = shallow(<Dropdown header="Menu header" items={items} />);
		expect(wrapper).toMatchSnapshot();
	});

	test('Sets pre-selected value', () => {
		const wrapper = shallow(<Dropdown header="Menu header" selectedItem={{ title: 'Ocean', id: 'item3' }} items={items} />);
		expect(wrapper.find('.dropdown-interactive-area').find('input').props().placeholder).toEqual('Ocean');
	});

	test('Searchable dropdown', () => {
		const wrapper = shallow(<Dropdown header="Menu header" searchable items={items} placeholder="Search and select" />);
		expect(wrapper.find('.dropdown-interactive-area').find('input').props().disabled).toEqual(false);
	});

	test('Change classname to open on click', () => {
		const wrapper = shallow(<Dropdown header="Menu header" items={items} />);
		wrapper.find('.dropdown-interactive-area').simulate('click');
		expect(wrapper.find('input').hasClass('focused')).toEqual(true);
		expect(wrapper.exists('.list-of-options')).toEqual(true);
	});

	test('Choose first item', () => {
		const wrapper = shallow(<Dropdown header="Menu header" items={items} open />);
		wrapper.find('.list-of-options').find('button').first().simulate('click');
		expect(wrapper.find('.dropdown-interactive-area').find('input').props().placeholder).toEqual('Apples');
	});

	test('Triggers filter function on search', () => {
		const wrapper = shallow(<Dropdown searchField items={items} open />);
		expect(wrapper.find('.option-list-element')).toHaveLength(3);
		wrapper.find('input').simulate('change', {target: {value: 'oc'}});
		expect(wrapper.find('.option-list-element')).toHaveLength(1);
	});

});
