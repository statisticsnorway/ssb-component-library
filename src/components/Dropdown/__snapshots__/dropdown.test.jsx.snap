// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component Matches the snapshot 1`] = `
<div
  className="ssb-dropdown"
>
  <label
    htmlFor={1}
  >
    Menu header
  </label>
  <button
    className="dropdown-interactive-area"
    onClick={[Function]}
  >
    <input
      className=""
      disabled={true}
      id={1}
      onChange={[Function]}
      placeholder="-- Select --"
      value=""
    />
    <ChevronDown
      className="dd-icon"
      color="currentColor"
      size={24}
    />
  </button>
</div>
`;
